openhab doesn't recognize squeezebox server at server boot

After booting my debian server all processes starts up. Especially squeezebox server (LMS) needs some minutes to come up. In general, openhab starts faster than LMS. I assume that this is the reason, why openhab does not recognize squeezebox server. Later, when squeezbox server has started up, openhab obviously doesn't care about. Just a reboot of openhab solves this and LMS is connected to OH.
I would expect, that the squieezebox binding look on regular basis after the squeezebox server.

I run OH 1.8.3 on a debain 8.5 server. Logitech Media Server Version: 7.9.0.

Is this a problem, which is known or do just I have it?
##

Not having a squeezebox, I can only test the "continuously fails to connect" case...

>22:21:58.170 [ERROR] [io.squeezeserver.SqueezeServer:523  ] - Failed to connect to the Squeeze Server at 127.0.0.1:9090
22:21:58.211 [INFO ] [io.squeezeserver.SqueezeServer:545  ] - Waiting 2 seconds before retrying failed connection.
22:22:00.213 [INFO ] [io.squeezeserver.SqueezeServer:551  ] - Retrying failed connection (attempt 1/5).
22:22:01.224 [ERROR] [io.squeezeserver.SqueezeServer:523  ] - Failed to connect to the Squeeze Server at 127.0.0.1:9090
...
22:22:13.573 [INFO ] [io.squeezeserver.SqueezeServer:545  ] - Waiting 2 seconds before retrying failed connection.
22:22:15.574 [INFO ] [io.squeezeserver.SqueezeServer:551  ] - Retrying failed connection (attempt 5/5).
22:22:16.584 [ERROR] [io.squeezeserver.SqueezeServer:523  ] - Failed to connect to the Squeeze Server at 127.0.0.1:9090
22:22:16.632 [ERROR] [io.squeezeserver.SqueezeServer:559  ] - All retries exhausted.  Giving up on connecting to server.

I've made the number of retries and the delay between them configurable, with defaults of 3 and 60, respectively.
>squeeze:server.retries=5
squeeze:server.retryTimeout=2

I've made [the jar for the updated squeeze server](https://drive.google.com/file/d/0B8vYheWfObP9VE16Y1psS1pLVEE/view?usp=sharing) available, if you'd like to give it a try.

I substituted my org.openhab.io.squeezeserver-1.8.0-SNAPSHOT.jar by org.openhab.io.squeezeserver-1.9.0-SNAPSHOT.jar in the addons folder, but following error makes my curious:

> 2016-09-17 19:04:20.935 [WARN ] [io.squeezeserver.SqueezeServer] - Unexpected or unsupported configuration: server.retries. Ignoring.
> 2016-09-17 19:04:20.936 [WARN ] [io.squeezeserver.SqueezeServer] - Unexpected or unsupported configuration: server.retryTimeout. Ignoring.
> 2016-09-17 19:04:20.971 [ERROR] [io.squeezeserver.SqueezeServer] - Failed to connect to the Squeeze Server at 192.168.2.8:9090
> java.net.ConnectException: Verbindungsaufbau abgelehnt
>         at java.net.PlainSocketImpl.socketConnect(Native Method) ~[na:1.8.0_101]
>         at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[na:1.8.0_101]
>         at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[na:1.8.0_101]
>         at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[na:1.8.0_101]
>         at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[na:1.8.0_101]
>         at java.net.Socket.connect(Socket.java:589) ~[na:1.8.0_101]
>         at java.net.Socket.connect(Socket.java:538) ~[na:1.8.0_101]
>         at java.net.Socket.<init>(Socket.java:434) ~[na:1.8.0_101]
>         at java.net.Socket.<init>(Socket.java:211) ~[na:1.8.0_101]
>         at org.openhab.io.squeezeserver.SqueezeServer.connect(SqueezeServer.java:488) [bundlefile:na]
>         at org.openhab.io.squeezeserver.SqueezeServer.updated(SqueezeServer.java:483) [bundlefile:na]
>         at org.eclipse.equinox.internal.cm.ManagedServiceTracker$1.run(ManagedServiceTracker.java:183) [org.eclipse.equinox.cm_1.0.400.v20120522-1841.jar:na]
>         at org.eclipse.equinox.internal.cm.SerializedTaskQueue$1.run(SerializedTaskQueue.java:36) [org.eclipse.equinox.cm_1.0.400.v20120522-1841.jar:na]

It seems, that the options

> squeeze:server.retries=5
> squeeze:server.retryTimeout=2

are not recognized. The running openHAB server does not reconnect to Squeezebox Server with this jar. Maybe you uploaded the wrong version?

--- a/bundles/io/org.openhab.io.squeezeserver/src/main/java/org/openhab/io/squeezeserver/SqueezeServer.java
+++ b/bundles/io/org.openhab.io.squeezeserver/src/main/java/org/openhab/io/squeezeserver/SqueezeServer.java
@@ -51,13 +51,15 @@ public class SqueezeServer implements ManagedService {
     // configuration defaults for optional properties
     private static final int DEFAULT_CLI_PORT = 9090;
     private static final int DEFAULT_WEB_PORT = 9000;
+    private static final int DEFAULT_RETRIES = 3;
+    private static final int DEFAULT_RETRY_TIMEOUT = 60;
 
     private static final String DEFAULT_TTS_URL = "http://translate.google.com/translate_tts?tl=en&ie=UTF-8&client=openhab&q=%s";
     private static final int DEFAULT_TTS_MAX_SENTENCE_LENGTH = 100;
 
     // / regEx to validate SqueezeServer config
     // <code>'^(squeeze:)(host|cliport|webport)=.+$'</code>
-    private static final Pattern SERVER_CONFIG_PATTERN = Pattern.compile("^(server)\\.(host|cliport|webport)$");
+    private static final Pattern SERVER_CONFIG_PATTERN = Pattern.compile("^(server)\\.(host|cliport|webport|retries|retryTimeout)$");
 
     // regEx to validate a mpdPlayer config <code>'^(.*?)\\.(id)$'</code>
     private static final Pattern PLAYER_CONFIG_PATTERN = Pattern.compile("^(.*?)\\.(id)$");
@@ -80,6 +82,8 @@ public class SqueezeServer implements ManagedService {
     private String host;
     private int cliPort;
     private int webPort;
+    private int retries;
+    private int retryTimeout;
 
     // client socket and listener thread
     private Socket clientSocket;
@@ -96,6 +100,8 @@ public class SqueezeServer implements ManagedService {
     // tts properties
     private String ttsUrl;
     private int ttsMaxSentenceLength;
+
+    private String _mutex = "wait";
 
     public synchronized boolean isConnected() {
         if (clientSocket == null) {
@@ -436,7 +442,8 @@ public class SqueezeServer implements ManagedService {
         host = null;
         cliPort = DEFAULT_CLI_PORT;
         webPort = DEFAULT_WEB_PORT;
-
+        retries = DEFAULT_RETRIES;
+        retryTimeout = DEFAULT_RETRY_TIMEOUT;
         ttsUrl = DEFAULT_TTS_URL;
         ttsMaxSentenceLength = DEFAULT_TTS_MAX_SENTENCE_LENGTH;
 
@@ -474,6 +481,10 @@ public class SqueezeServer implements ManagedService {
                     cliPort = Integer.valueOf(value);
                 } else if (serverConfig.equals("webport") && StringUtils.isNotBlank(value)) {
                     webPort = Integer.valueOf(value);
+                } else if (serverConfig.equals("retries") && StringUtils.isNotBlank(value)) {
+                    retries = Integer.valueOf(value);
+                } else if (serverConfig.equals("retryTimeout") && StringUtils.isNotBlank(value)) {
+                    retryTimeout = Integer.valueOf(value);
                 }
             } else if (playerMatcher.matches()) {
                 String playerId = playerMatcher.group(1);
@@ -487,7 +498,7 @@ public class SqueezeServer implements ManagedService {
             } else if (ttsMaxSentenceLengthMatcher.matches() && StringUtils.isNotBlank(value)) {
                 ttsMaxSentenceLength = Integer.valueOf(value);
             } else {
-                logger.warn("Unexpected or unsupported configuration: " + key + ". Ignoring.");
+                logger.warn("Ignored unexpected or unsupported configuration: {}", key);
             }
         }
 
@@ -500,15 +511,17 @@ public class SqueezeServer implements ManagedService {
         }
 
         // attempt to connect using our new config
-        connect();
+        if (!connect()) {
+            retryConnect();
+        }
     }
 
-    private void connect() {
+    private boolean connect() {
         try {
             clientSocket = new Socket(host, cliPort);
         } catch (IOException e) {
             logger.error("Failed to connect to the Squeeze Server at " + host + ":" + cliPort, e);
-            return;
+            return false;
         }
 
         try {
@@ -516,10 +529,35 @@ public class SqueezeServer implements ManagedService {
             listener.start();
         } catch (IllegalThreadStateException e) {
             logger.error("Failed to start the Squeeze Server listener thread", e);
-            return;
+            return false;
         }
 
         logger.info("Squeeze Server connection started.");
+        return true;
+    }
+
+    private boolean retryConnect() {
+        int retriesRemaining = retries;
+        boolean lastResult = false;
+
+        while (retriesRemaining > 0) {
+            synchronized (_mutex) {
+                logger.info("Waiting {} seconds before retrying failed connection.", retryTimeout);
+                try {
+                    _mutex.wait(retryTimeout*1000);
+                }catch(InterruptedException e) {
+                    logger.warn("Timeout was interrupted.");
+                 }
+                logger.info("Retrying failed connection (attempt {}/{}).", retries-retriesRemaining+1, retries);
+                if (connect()) {
+                    return true;
+                }
+                retriesRemaining--;
+            }
+        }
+
+        logger.error("All retries exhausted.  Giving up on connecting to server.");
+        return false;
     }
 
     private void disconnect() {

--- a/distribution/openhabhome/configurations/openhab_default.cfg
+++ b/distribution/openhabhome/configurations/openhab_default.cfg
@@ -1308,6 +1308,13 @@ hue:secret=
 # Host (IP address) of your Logitech Media Server
 #squeeze:server.host=
 
+# Number of retries to allow for a failed connection. (optional; defaults to 3)
+#squeeze:server.retries=
+
+# Timeout (in seconds) between retries of a failed connection to the
+# Logitech Media Server. (optional; defaults to 60)
+#squeeze:server.retryTimeout=
+
 # Port of CLI interface of your Logitech Media Server (optional, defaults to 9090)
 #squeeze:server.cliport=
 

